docs/tutorials-redeliver-messages.md

```diff
@@ -0,0 +1,236 @@
+In many message queue systems (such as RabbitMQ, Google Pub/Sub, Apache Pulsar, etc.), this strategy is implemented.
+In Pulsar, the deadLetterPolicy is implemented at the client side, it creates a new retry letter topic and dead letter
+topic when building a consumer with `deadLetterPolicy` configuration. When a consumer calls `reconsumeLaterCumulative`
+or `reconsumeLater`, the message (method parameter) will be produced to the retry letter topic until the retry time reaches
+the `maxRedeliverCount`. The message will be produced to the dead letter topic when the retry time reaches the `
+maxRedeliverCount`. The main difference between them is that `reconsumeLaterCumulative` will cumulative ack the
+message (method parameter) after it is produced and `reconsumeLater` will individual ack the message (method parameter)
+after it is produced.
+
+The `negativeAcknowledge` is used to redeliver certain unacknowledged messages while `redeliverUnacknowledgedMessages`
+is used to redeliver all the unacknowledged messages received by this consumer. The main difference between them and
+deadLetterPolicy is that there is no new topic created, and there is an unlimited number of redeliveries.
+
+## Best Practice Suggestion
+
+Different scenarios require different best practices. Users who value the order of partition messages and wish to batch
+process data should choose or implement an appropriate routerPolicy to send a batch of ordered messages to the same
+partition. They should also select either Exclusive or Failover subscription modes. For users who do not care about
+message order and those in stream processing scenarios, they can opt to use Shared and Key-shared subscription modes.
+
+### Shared && Key-shared
+#### At-least-once
+The term `At least once` ensures that the message is processed at least once, and there is a risk of duplicate messages,
+but the performance is better than the `Exactly once` semantics. For the `At least once` semantics in shared or
+key-shared mode, the most important matter is avoiding ack-hole as much as possible. The ack-hole refers to instances
+where some single messages are missed for acknowledgment. This can occur in the following cases (All of these are the cases
+where the connection is not interrupted, and reconnection will automatically resend the message.):
+
+1. Consumer receives messages but fails to process due to business system error.
+2. Consumer receives messages but misses to process or the business system gets stuck.
+3. Consumer acknowledges the message, but the acknowledge request is lost when sending to the broker. The possibility of
+4. data loss exists in the case of TCP long connections, although the probability is extremely low.
+   ![img_8.png](../static/img/blog-consume-best-practice/Ack-hole.png)
+
+For case 1, configuring `deadLetterPolicy` is a good solution. When the business system receives a signal of message
```

Comment:
This explanation is redundant IMO. Let's expand at the section above.

Reply:
Ok, I think you are correcting the structure of this article. 
The original article is like the following:
```
# Background knowledge
   ## subscription type
   ## acknowledgment 
   ## redeliver method
# best practice 
  ## shared/ key shared
       ### At-least-once
      ### Exactly-once
 ## Exclusive && Failover
      ### At-least-once
     ### Exactly-once
```
And you want to delete the part of best practice and expand it in the part of the redeliver method, right?